discovery_engine/lib/src/ffi/types/result.dart

```diff
@@ -24,132 +24,161 @@ import 'package:xayn_discovery_engine/src/ffi/types/primitives.dart'
     show Uint8ListFfi;
 import 'package:xayn_discovery_engine/src/ffi/types/string.dart' show StringFfi;
 
-class ResultFfiAdapter<Ok, Err, RustResult extends NativeType,
+abstract class ResultFfiAdapter<Ok, Err, RustResult extends NativeType,
     RustOk extends NativeType, RustErr extends NativeType> {
-  final Pointer<RustOk> Function(Pointer<RustResult>) getOk;
-  final Pointer<RustErr> Function(Pointer<RustResult>) getErr;
-  final Pointer<RustOk> Function(Pointer<RustResult>) moveOk;
-  final Pointer<RustErr> Function(Pointer<RustResult>) moveErr;
-  final Ok Function(Pointer<RustOk>) readNativeOk;
-  final Err Function(Pointer<RustErr>) readNativeErr;
-  final Never Function(Err) throwErr;
-  final void Function(Pointer<RustOk>) freeOk;
-  final void Function(Pointer<RustErr>) freeErr;
-  final void Function(Pointer<RustResult>) freeResult;
-
-  ResultFfiAdapter({
-    required this.getOk,
-    required this.getErr,
-    required this.moveOk,
-    required this.moveErr,
-    required this.readNativeOk,
-    required this.readNativeErr,
-    required this.throwErr,
-    required this.freeOk,
-    required this.freeErr,
-    required this.freeResult,
-  });
+  late final Pointer<RustOk> Function(Pointer<RustResult>) _getOk;
+  late final Pointer<RustErr> Function(Pointer<RustResult>) _getErr;
+  late final Ok Function(Pointer<RustOk>) _readNativeOk;
+  late final Err Function(Pointer<RustErr>) _readNativeErr;
+  late final Never Function(Err) _throwErr;
+  late final void Function(Pointer<RustResult>) _freeResult;
```

Comment:
Why are they lazy internalized?

Reply:
because the field isn't initialized in the abstract class/mixin, but only in the subclass later on. even though it gets immediately initialized in the subclass constructor, dart doesn't allow this field to not be lazy (i guess the superclass can in general not know what the subclass is going to do with the field).

Reply:
yes, but you could just have a constructor in the super-class and then call that in the initializer list in the sub-class.

Reply:
that would only remove the `late` from the base class, the mixins still have to use `late`.